### PR TITLE
TLS and mTLS support

### PR DESCRIPTION
**Pull request by macrogreg**
### Summary

Changes required to support TLS/SSL and mutual TLS connections on all supported .NET versions.

### Related work items

* https://github.com/temporalio/sdk-dotnet/issues/23
* https://github.com/temporalio/sdk-dotnet/issues/24
* https://github.com/temporalio/sdk-dotnet/issues/35

#### Tests & Validations
We continue to use manual validation for the functionality.
Automated unit and integration testing is getting ever closer, but we still need to implement a few bits and pieces to provide this effectively. In particular, integration testing will be easier to implement once serialization functionality is complete.

#### Feature limitations

This PR enables .NET workflow client support for TLS and mTLS on all supported .NET versions.
Note that different low-level gRPC libraries are used on different .NET versions:
 1. On Net Fx we use a gPRC Core C-Lib based library (https://grpc.io/docs/languages/csharp/dotnet/)
 2. On Net Core and Net modern we use gRPC.Net (https://github.com/grpc/grpc-dotnet)

(This is standard.)

The legacy gPRC Core C-Lib based library has some limitations around TLS support. In this version, we choose not to provide solutions for those limitations, but to expose them to the user. This allows us to enable this support and move on to adding other features. If something different becomes a popular user request, we will address it accordingly.
For now, this means 2 limitations for the classic Net Fx:
    a) `SkipServerCertValidation` is not supported. Skipping server validation may be required when you are using self-signed server certificates in test scenarios, and you do not want to distribute your custom test CA certificates onto your test client fleet. On implementation level, skipping server cert validation requires custom code to run during sever certificate verification. The gPRC Core C-Lib based library only allows such custom code to run _after_ a successful server cert validation. To work around this limitation, distribute the custom CA certs on your clients.
    b) Certificate (and key) file/data format. On Net FX we only support PEM format. On Net Core and Net modern we support both, PEM and PFX ([X509Certificate2](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2)-based) certificates with Windows Cert Store support. This is also due to a gPRC Core C-Lib limitation.

### CODE REVIEW

File: Src/SDK/WorkflowClient/Temporal.WorkflowClient/public/TemporalClient.cs [hunk 1]
```diff
@@ -412,6 +414,66 @@ private List<ITemporalClientInterceptor> CreateDefaultServiceInvocationPipeline(
...
+                    // @ToDo: once we have logging, log if on finalizer thread.
+                    ExceptionDispatchInfo.Capture(ex).Throw();
+                }
+            }
+        }
+
+        ~TemporalClient()
+        {
+            // Do not change this code. Put cleanup code in `Dispose(bool disposing)` method.
+            Dispose(disposing: false);
```
Comment (Spikhalskiy) [inline Src/SDK/WorkflowClient/Temporal.WorkflowClient/public/TemporalClient.cs:467]: Isn't it a default implementation of the finalizer? Why do we need it here?

File: Src/SDK/WorkflowClient/Temporal.WorkflowClient/public/TemporalClientConfiguration.Connection.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                                                                  isTlsEnabled: true);
+            }
+
+            /// <summary>
+            /// </summary>
+            /// <remarks>
+            /// This overload is only available on Net Core and Net 5+. See <see cref="TemporalClientConfiguration.TlsCertificate"/>
+            /// for details on this topic.
+            /// </remarks>
+            public static TemporalClientConfiguration.Connection ForTemporalCloud(string @namespace,
```
Comment (Spikhalskiy) [inline Src/SDK/WorkflowClient/Temporal.WorkflowClient/public/TemporalClientConfiguration.Connection.cs:61]: Let's not have `ForTemporalCloud` in public APIs. Temporal Cloud is even not GA.
Let's put this code somewhere in tests for now if it's needed there.

File: Src/SDK/WorkflowClient/Temporal.WorkflowClient/internal/WorkflowServiceClientEnvelope.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+{
+    internal sealed class WorkflowServiceClientEnvelope
+    {
+        private readonly WorkflowService.WorkflowServiceClient _grpcWorkflowServiceClient;
+        private readonly ChannelBase _grpcChannel;
+        private readonly TemporalClientConfiguration.Connection _connectionConfig;
+
+        private readonly object _refCountLock = new();  // Held for very short periods and relatively seldom.
+
+        private int _refCount = 0;
```
Comment (Spikhalskiy) [inline Src/SDK/WorkflowClient/Temporal.WorkflowClient/internal/WorkflowServiceClientEnvelope.cs:21]: Can you add a short javadoc about the purpose of reference counting here?

**Review by Spikhalskiy — APPROVED** [review]

File: Src/SDK/WorkflowClient/Temporal.WorkflowClient/internal/WorkflowServiceClientEnvelope.cs (hunk 3 above)
Comment (macrogreg) [reply Src/SDK/WorkflowClient/Temporal.WorkflowClient/internal/WorkflowServiceClientEnvelope.cs:23]: Good idea. Done.
https://github.com/temporalio/sdk-dotnet/pull/44/commits/360e1864474777d7ae94718c87e0255dafdce082

File: Src/SDK/WorkflowClient/Temporal.WorkflowClient/public/TemporalClientConfiguration.Connection.cs (hunk 2 above)
Comment (macrogreg) [reply Src/SDK/WorkflowClient/Temporal.WorkflowClient/public/TemporalClientConfiguration.Connection.cs:61]: I had some feedback from the Cloud team that they were very glad about this feature. Unfortunately I cannot remember who exactly said it. So I created a tracking item to make suer this does not get forgotten, and leaving it for now. We'll sync up with the cloud team and decide on the best way around this.
Tracker: https://github.com/temporalio/sdk-dotnet/issues/54

**Pull request merged**